File: app/src/main/java/com/travlog/android/apps/ui/adapters/NoteAdapter.java
package com.travlog.android.apps.ui.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.travlog.android.apps.R;
import com.travlog.android.apps.models.Note;
import com.travlog.android.apps.ui.viewholders.NoteViewHolder;

import java.util.ArrayList;
import java.util.List;

public class NoteAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    public interface Delegate extends NoteViewHolder.Delegate {
    }

    private final Delegate delegate;
    private final List<Note> notes;
    private final SparseArray<Note> noteSparseArray;

    public NoteAdapter(final @NonNull Delegate delegate) {
        this.delegate = delegate;
        this.notes = new ArrayList<>();
        this.noteSparseArray = new SparseArray<>();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new NoteViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.i_note, parent, false), delegate);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ((NoteViewHolder) holder).bindData(notes.get(position));
    }

    @Override
    public int getItemCount() {
        return notes.size();
    }

    public boolean updateData(final @NonNull List<Note> notes) {
        final boolean result = this.notes.addAll(notes);

        for (Note note : notes) {
            noteSparseArray.put(note.id, note);
        }

        notifyDataSetChanged();
        return result;
    }

    public @NonNull
    Note deleteData(final int noteId) {
        final Note note = noteSparseArray.get(noteId);
        final int index = notes.indexOf(note);

        notes.remove(index);
        noteSparseArray.delete(noteId);
        notifyItemRemoved(index);

        return note;
    }
}

File: app/src/main/java/com/travlog/android/apps/ui/adapters/NoteAdapter.kt
/*
 * Copyright 2019 Travlog. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.travlog.android.apps.ui.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.travlog.android.apps.R
import com.travlog.android.apps.models.Note
import com.travlog.android.apps.ui.viewholders.NoteViewHolder
import timber.log.Timber
import java.util.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.set

class NoteAdapter(private val delegate: Delegate) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val notes: MutableList<Note>
    private val noteMap: HashMap<String, Note>

    interface Delegate : NoteViewHolder.Delegate

    init {
        this.notes = ArrayList()
        this.noteMap = HashMap()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            NoteViewHolder(LayoutInflater.from(parent.context)
                    .inflate(R.layout.i_note, parent, false), delegate)

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) =
            (holder as NoteViewHolder).bindData(notes[position])

    override fun getItemCount(): Int = notes.size

    fun clearData() =
            notes.clear().run {
                Timber.d("clearData")
                noteMap.clear()
                notifyDataSetChanged()
            }

    fun updateData(notes: List<Note>): Boolean {
        Timber.d("updateData? %s", notes)
        this.notes.addAll(notes).let {
            for (note in notes) {
                noteMap[note.id] = note
            }

            notifyDataSetChanged()

            return it
        }
    }

    fun updateNote(note: Note) {
        Timber.d("updateNotes? %s", note)
        val originNote = noteMap.get(note.id)

        noteMap[note.id] = note

        if (originNote != null) {
            val index = notes.indexOf(originNote)

            notes[index] = note
            notifyItemChanged(index)
        } else {
            notes.add(0, note)
            notifyItemInserted(0)
        }
    }

    fun deleteData(nid: String): Note? {
        val note = noteMap[nid]
        val index = notes.indexOf(note)

        if (index >= 0) {
            notes.removeAt(index)
            noteMap.remove(nid)
            notifyItemRemoved(index)
        }

        return note
    }
}
